### ISSUE TITLE
[3.3] Can't get template listing on themes with "links"

### ISSUE THREAD

**Issue by bobdenotter** [bolt/bolt#6456]
If your theme has a node folder, there's a good chance it has symlinks in it. Finder doesn't like this. See screenshot: 

![screen shot 2017-03-04 at 15 18 20](https://cloud.githubusercontent.com/assets/1833361/23579440/d42d955a-00ed-11e7-83e8-7f709ced65ea.png)

It breaks in `public function listTemplates`, which is weird, because it has this: 

```

        /** @var Finder|FileInterface[] $finder */
        $finder = $this->templatesDir->find()
            ->files()
            ->notName('/^_/')
            ->notPath('node_modules')
            ->notPath('bower_components')
            ->notPath('.sass-cache')
            ->depth('<4')
            ->path($name)
            ->sortByName()
        ;
```

So, `node_modules` should be ignored altogether, but it doesn't/

Details
-------

 - Relevant Bolt Version: [ 3.3 | master ]

**Comment by bobdenotter** [comment]
@CarsonF This seems to be something in `\Bolt\Filesystem\Finder`, so assigning to you. :-)

**Comment by GawainLynch** [comment]
To re-re-re-re-re-re-re-re-re-repeat. The exception is thrown by [Flysystem](https://github.com/thephpleague/flysystem/blob/1.0.35/src/Adapter/Local.php#L429) as while there is an exclude, PHP still scans everything first.

As has been [linked here before](https://github.com/thephpleague/flysystem/issues/474#issuecomment-97453514):

> A fix has been pushed to make symlinks explicitly unsupported. This is done because they can't be handled without forcing a particular strategy upon the user. Flysystem respects a root, symlinks are possible leaks. A resolved symlink could be outside of the defined root. In that case Flysystem can't handle it, but can't ignore it either. If a symlink is used to keep files outside of a webroot (accessible to the public) a symlinks should be placed on the root directory of the adapter. This is supported.

**Comment by bobdenotter** [comment]
I know it's a flysystem thing, but that doesn't mean it's not "broken" from the perspective of the user. This is something that needs fixing, imho.

**Comment by GawainLynch** [comment]
Then we remove RFS support.

**Comment by bobdenotter** [comment]
Let's discuss this in the dev-meeting.

**Comment by GawainLynch** [comment]
I'll be somewhere about 11 kilometres above South East Asia at that point

**Comment by bobdenotter** [comment]
Then we discuss it at the _next_ dev-meeting :-)

**Comment by GawainLynch** [comment]
I *won't* be somewhere about 11 kilometres above South East Asia at that point :stuck_out_tongue_winking_eye:

**Comment by bobdenotter** [comment]
I was sort of assuming that you'd be at that exact spot for less than a full week, yes. 😎

**Comment by CarsonF** [comment]
@bobdenotter Try changing the `notPath()` calls to `exclude()`. That _should_ work and be more performant too. 

The difference being `notPath()` gathers all the files _first_ then filters out the paths in `notPath`, but `exclude()` doesn't even descend into directories that are in `exclude()`.

The only thing I'm unsure of is if the path of the directory matters. i.e. `./node_modules` vs `./foo/node_modules`. I have a feeling it's the former.

**Comment by bobdenotter** [comment]
@CarsonF Thank you, mate! That was exactly what was needed. PR incoming.

**Comment by bobdenotter** [comment]
Fixed in #6460
